### PR TITLE
Add a `container_context` field to RepositoryPythonOrigin that can be used for additional container-specific metadata

### PR DESCRIPTION
**Pull request by gibsondan**
Summary:
This dict then ends up on the run, which can then be accessed within the run launcher and the executor to incorporate per-location config into the resulting containers.

## Checklist

- [ ] My change requires a change to the documentation and I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.

### CODE REVIEW

File: python_modules/libraries/dagster-docker/dagster_docker/container_context.py [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+        )
+
+    def merge(self, other: "DockerContainerContext"):
+        return DockerContainerContext(
+            registry=other.registry if other.registry != None else self.registry,
+            env_vars=self.env_vars + other.env_vars,
+            networks=self.networks + other.networks,
+            container_kwargs=other.container_kwargs
+            if other.container_kwargs != None
+            else self.container_kwargs,
```
Comment (gibsondan) [inline python_modules/libraries/dagster-docker/dagster_docker/container_context.py:51]: this part merits some discussion probably. Note how some fields (the lists, basically) are appended, while others are replaced. This mirrors the existing logic for transferring run launcher config onto executors (and is very likely what you want for things like secret) - but might be confusing.

File: python_modules/libraries/dagster-docker/dagster_docker/docker_run_launcher.py [hunk 2]
```diff
@@ -64,13 +66,22 @@ def config_type(cls):
     def from_config_value(inst_data, config_value):
         return DockerRunLauncher(inst_data=inst_data, **config_value)
 
-    def _get_client(self):
+    @property
+    def instance_container_context(self) -> DockerContainerContext:
+        return DockerContainerContext(
+            self.registry, self.env_vars, self.networks, self.container_kwargs
+        )
```
Comment (alangenfeld) [inline python_modules/libraries/dagster-docker/dagster_docker/docker_run_launcher.py:73]: whats this? the `instance_` prefix is confusing given conceptual gravity of DagsterInstance

File: python_modules/libraries/dagster-docker/dagster_docker/docker_executor.py [hunk 3]
```diff
@@ -63,117 +64,133 @@ def docker_executor(init_context: InitExecutorContext) -> Executor:
...
         )
 
     def launch_step(self, step_handler_context: StepHandlerContext) -> List[DagsterEvent]:
-        client = self._get_client()
-
-        step_image = (
-            step_handler_context.execute_step_args.pipeline_origin.repository_origin.container_image
-        )
+        # TODO Would be nice to have access to the run before each individual launch_step
+        # call since the deployment target does not change
```
Comment (alangenfeld) [inline python_modules/libraries/dagster-docker/dagster_docker/docker_executor.py:174]: is this a trivial addition to `StepHandlerContext`  ?

File: python_modules/libraries/dagster-docker/dagster_docker/container_context.py [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+from dagster.core.errors import DagsterInvalidConfigError
+from dagster.core.storage.pipeline_run import PipelineRun
+
+from .utils import DOCKER_CONTAINER_CONTEXT_SCHEMA
+
+if TYPE_CHECKING:
+    from . import DockerRunLauncher
+
+
+class DockerContainerContext(
```
Comment (alangenfeld) [inline python_modules/libraries/dagster-docker/dagster_docker/container_context.py:14]: docblock somewhere for this  - mostly just to capture current intent behind object

**Review by alangenfeld — COMMENTED** [review]
The biggest concern i have around the `merge` behavior is it being a breaking change for users if we realize there is something smarter. Any clever ideas around that?

**Review by johannkm — DISMISSED** [review]
surfacing the comments

File: python_modules/dagster-test/dagster_test/test_project/__init__.py [hunk 5]
```diff
@@ -114,12 +114,13 @@ def build_and_tag_test_image(tag):
     return subprocess.check_output(["./build.sh", base_python, tag], cwd=get_test_repo_path())
 
 
-def get_test_project_recon_pipeline(pipeline_name, container_image=None):
+def get_test_project_recon_pipeline(pipeline_name, container_image=None, container_context=None):
```
[the remark this answers is not in the document]
Comment (alangenfeld) [reply python_modules/dagster-test/dagster_test/test_project/__init__.py:117]: I'm not entirely convinced on the subprocess env var case due to how env var inheritance works. I think if you are programmatically setting up environment, its very likely happening at the layer where you are starting the container.

`container_context` feels very much an extension of `container_image`, and in the future may consolidate together into something more coherent

**Review by alangenfeld — DISMISSED** [review]

**Comment by gibsondan** [comment]
Current dependencies on/for this PR:
* master
  * **PR #6925** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/6925" target="_blank">Graphite</a>  (this PR)
    * **PR #7176** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/7176" target="_blank">Graphite</a> 
      * **PR #7203** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/7203" target="_blank">Graphite</a> 

This comment was auto-generated by [Graphite](https://graphite.dev?utm_source=gh).

**Review by clairelin135 — APPROVED** [review]

**Comment by clairelin135** [comment]
I don't know why or how I dismissed the review above, but approving now

**Pull request merged**